spices/SPICE-0001-import-graph-analyzer-api.adoc

```diff
@@ -0,0 +1,336 @@
+----
+
+This command parses the provided source modules, and resolves all imported modules to their absolute URI.
+
+By default, the result is Pcf. Example:
+
+[source,shell]
+----
+pkl analyze imports myFile.pkl
+----
+
+Produces:
+[source]
+----
+imports {
+  ["file:///path/to/myFile.pkl"] {
+    "file:///path/to/otherFile.pkl"
+  }
+  ["file:///path/to/otherFile.pkl"] {}
+}
+resolvedUris {
+  ["file:///path/to/myFile.pkl"] = "file:///path/to/myFile.pkl"
+  ["file:///path/to/myFile.pkl"] = "file:///path/to/myFile.pkl"
+}
+----
+
+The value of the `imports` property is an object whose keys are module URIs, and whose values are the set of modules imported by the module referenced by the key.
+
+The set of all input modules in a program is the set of keys in the `imports` object.
+
+The value of `resolvedUris` is the mapping of in-language module URIs to those that Pkl uses to determine how to fetch the module's contents.
+For example, in the case of a link:https://pkl-lang.org/main/current/language-reference/index.html#local-dependencies[local dependency], an in-language URI might look like `projectpackage://example.com/foo@1.0.0#/mod.pkl`, and its resolved URI might look like `\file:///path/to/foo/mod.pkl`.
+
+The resolved URIs help build systems understand how resources map
+
+Underneath the hood, the command builds a synthetic module, that ultimately uses standard library module <<stdlib-module-pkl-analyze,pkl.analyze>> to produce output.
+
+==== `--format`, `-f`
+
+The `--format` (`-f` for short) flag sets the output format for this command.
+
+Examples:
+
+[source,shellscript]
+----
+# Evaluate as JSON
+$ pkl analyze imports -f json myFile.pkl
+
+# Evaluate as YAML
+$ pkl analyze imports -f yaml myFile.pkl
+----
+
+==== Common CLI flags
+
+This command recursively traverses the import graph.
+To support this, the set of flags from common evaluator settings are available:
+
+* `--allowed-modules`
+* `--allowed-resources`
+* `--cache-dir`
+* `--module-path`
+* `--project-dir`
+* `--ca-certificates`
+
+[[stdlib-module-pkl-analyze]]
+=== Standard library module `pkl:analyze`
+
+A new standard library module is introduced.
+It contains the class `ImportAnalyzer`, which provides import graph analysis.
+
+[source,pkl]
+----
+/// Utilities for static analysis of Pkl modules.
+///
+/// Unlike `pkl:reflect`, these methods do not require loading the modules for analysis.
+@ModuleInfo { minPklVersion = "0.26.0" }
+@Since { version = "0.26.0" }
+module pkl.analyze
+
+class ImportAnalyzer {
+  /// The modules to analyze for imports.
+  moduleUris: Listing<String>
+
+  /// The graph of imports declared by [moduleUris].
+  /// Each key-value pair is a module, corresponding with the set of modules
+  /// imported by that module.
+  ///
+  /// The set of all modules in the graph can be obtained via its keys.
+  external imports: Mapping<Uri, Listing<Uri>(isDistinct)>
```

Comment:
Wouldn't `Map` be more efficient/capable?

Reply:
Yeah, perhaps. I went with `Mapping` because this is designed to produce rendered output.

This:

```
imports {
  ["file:///foo.pkl"] {
     "file:///bar1.pkl"
     "file:///bar2.pkl"
     "file:///bar3.pkl"
  }
}
```

Instead of:

```
imports = Map("file:///foo.pkl", List("file:///bar1.pkl", "file:///bar2.pkl", "file:///bar3.pkl"))
```

Reply:
Perhaps the rendering could use a converter?

Reply:
Fair!